src/subscribe.rs

```diff
@@ -0,0 +1,429 @@
+                );
+        }
+    }
+
+    user_data
+}
+
+fn confirmation_link(
+    key: &str,
+    email: &str,
+    ft_sdk::Host(host): &ft_sdk::Host,
+    ft_sdk::Mountpoint(mountpoint): &ft_sdk::Mountpoint,
+) -> String {
+    format!(
+        "https://{host}{mountpoint}{confirm_sub_route}?code={key}&email={email}",
+        confirm_sub_route = "/confirm-subscription/",
+        mountpoint = mountpoint.trim_end_matches('/'),
+    )
+}
+
+fn send_double_opt_in_email(
+    conn: &mut ft_sdk::Connection,
+    to: (&str, &str),
+    conf_link: &str,
+    topic: Option<String>,
+) -> Result<(), ft_sdk::Error> {
+    let (from_name, from_email) = email_from_address_from_env();
+
+    let name_or_email = if to.0.is_empty() { to.1 } else { to.0 };
+
+    let to_topic = if topic.is_some() {
+        format!("to the {}", topic.unwrap())
+    } else {
+        "".to_string()
+    };
+
+    let body_html = subscription::email_templ::CONFIRM_SUBSCRIPTION_EMAIL_TEMPLATE_HTML
+        .replace("{name}", name_or_email)
+        .replace("{confirmation_link}", conf_link)
+        .replace("{to_topic}", &to_topic);
+
+    let body_txt = subscription::email_templ::CONFIRM_SUBSCRIPTION_EMAIL_TEMPLATE_TXT
+        .replace("{name}", name_or_email)
+        .replace("{confirmation_link}", conf_link)
+        .replace("{to_topic}", &to_topic);
+
+    Ok(ft_sdk::send_email(
+        conn,
+        (&from_name, &from_email),
+        vec![to],
+        "Confirm your subscription",
+        &body_html,
+        &body_txt,
+        None,
+        None,
+        None,
+        "auth_confirm_account_request",
+    )?)
+}
+
+/// return `id` if it is Some
+/// otherwise, get user data by email and return its id
+/// if user data is not found, create user and return its id
+fn get_or_create_user_id(
+    conn: &mut ft_sdk::Connection,
+    id: Option<ft_sdk::auth::UserId>,
+    subscriber: &Subscriber,
+) -> Result<ft_sdk::auth::UserId, ft_sdk::Error> {
+    let user_id = if let Some(id) = id {
+        id
+    } else {
+        match ft_sdk::auth::provider::user_data_by_email(
+            conn,
+            crate::EMAIL_PROVIDER_ID,
+            &subscriber.email,
+        )
+        .map(|(id, _)| id)
+        {
+            Ok(v) => v,
+            Err(e) => match e {
+                ft_sdk::auth::UserDataError::NoDataFound => create_user(conn, subscriber)?,
+                e => return Err(e.into()),
+            },
+        }
+    };
+
+    Ok(user_id)
+}
+
+fn create_user(
+    conn: &mut ft_sdk::SqliteConnection,
+    subscriber: &Subscriber,
+) -> Result<ft_sdk::UserId, ft_sdk::Error> {
+    use diesel::prelude::*;
+    use ft_sdk::auth::fastn_user;
+
+    let data = serde_json::to_string(&serde_json::json!({
+        "email": {
+            "emails": [subscriber.email],
+            "identity": subscriber.email,
+        },
+        "subscription": {
+            "identity": subscriber.email,
+            "name": subscriber.name,
+        },
+    }))?;
```

Comment:
This is subscription app. It should not add other provider data. Since we should use `ft_sdk::auth::ProviderData` for all providers, other providers (like email auth) should take the leverage to get the required data.

Reply:
Also use `ft_sdk::auth::provider::create_user`

Reply:
We have not introduced the concept of `ft_sdk::AppData` yet. This code will change when that exists, we don't even know if we need that.

`ft_sdk::auth::provider::create_user` can't insert user data of this kind where multiple providers are set. This is wanted as the assumption that email should always be fetched from the `email` provider is made throughout the implementation of subscription and features in the [auth wasm](https://github.com/fastn-community/auth/) package